File: Week 1/stackInterface.go
package main

import (
  "fmt"
  "strings"
  "strconv"

  // For user input:
  "bufio"
  "os"
)

func processCommands(commandList []string) {
  var stack, maxes []int

  for _, v := range commandList {
    c := strings.Split(v, " ")
    if c[0] == "add" {
      convert, _ := strconv.Atoi(c[1])
      stack = append(stack, convert)
      if len(maxes) == 0 || maxes[len(maxes) - 1] < convert {
        maxes = append(maxes, convert)
      }
    } else if c[0] == "pop" {
      if stack[len(stack) - 1] == maxes[len(maxes) - 1] {
        maxes = maxes[:len(maxes) - 1]
      }
      stack = stack[:len(maxes) - 1]
    } else if c[0] == "max" {
      fmt.Println(maxes[len(maxes) - 1])
    }
  }
}

func main() {
  var commands []string
  var numCommands int
  fmt.Scanln(&numCommands)
  scanner := bufio.NewScanner(os.Stdin)

  for i := 0; i < numCommands; i++ {
    scanner.Scan()
    commands = append(commands, scanner.Text())
  }
}

File: Week 1/Ruby Solutions/stackInterface.rb
numCommands = gets.chomp.to_i

commands = []
numCommands.times do |i|
  command = gets.chomp.split(/ /)
  commands << command
end

def processCommands(commandList)
  stack, maxes = [], []
  commandList.each do |i|
    if i[0] == "push"
      if maxes.empty? || i[1].to_i > maxes.last then maxes.push(i[1].to_i) end
      stack.push(i[1].to_i)
    elsif i[0] == "pop"
      if i[1].to_i == maxes.last then maxes.pop end
      stack.pop
    elsif i[0] == "max"
      puts maxes.last
    end
  end
end

processCommands(commands)
